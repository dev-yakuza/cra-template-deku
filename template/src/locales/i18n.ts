import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

import en from './translation/en.json'
import ja from './translation/ja.json'
import ko from './translation/ko.json'

const defaultLanguage = 'en'
const languages = [
  {
    title: 'English',
    value: 'en',
  },
  {
    title: '日本語',
    value: 'ja',
  },
  {
    title: '한국어',
    value: 'ko',
  },
]

const resources = {
  en: {
    translation: en,
  },
  ja: {
    translation: ja,
  },
  ko: {
    translation: ko,
  },
}

i18n.use(initReactI18next).init({
  resources,
  lng: defaultLanguage,
  fallbackLng: defaultLanguage,
  interpolation: {
    escapeValue: false,
  },
})

export default i18n
export { defaultLanguage, languages }
