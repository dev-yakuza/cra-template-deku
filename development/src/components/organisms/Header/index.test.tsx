import { useEffect } from 'react'

import { act, fireEvent, screen } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { languages } from 'locales/i18n'
import { useTranslation } from 'react-i18next'
import { BrowserRouter, Router } from 'react-router-dom'
import { mockAppBar, mockIconSelectBox, render } from 'utils/test'

import { Header } from '.'

describe('<Header />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>,
    )

    const appBar = mockAppBar.mock.calls[0][0]
    expect(appBar.position).toBe('fixed')
    expect(appBar.elevation).toBe(4)
    expect(appBar.style).toEqual({
      backgroundColor: '#FFF',
    })

    const toolbar = appBar.children
    expect(toolbar.type.name).toBe('Toolbar')

    const children = toolbar.props.children
    {
      const box = children[0]
      expect(box.type.render.name).toBe('Box')
      expect(box.props.sx).toEqual({ flexGrow: 1 })

      const appTitleLink = box.props.children
      expect(appTitleLink.type.render.displayName).toBe('Styled(Link)')
      expect(appTitleLink.props.to).toBe('/')

      const appTitle = appTitleLink.props.children
      expect(appTitle.type.name).toBe('Typography')
      expect(appTitle.props.variant).toBe('h6')
      expect(appTitle.props.color).toBe('inherit')
      expect(appTitle.props.noWrap).toBe(true)
      expect(appTitle.props.children).toBe('Blog App')
    }

    {
      const iconSelectBox = children[1]
      expect(iconSelectBox.type.name).toBe('IconSelectBox')
      expect(iconSelectBox.props.icon.type.type.render.displayName).toBe(
        'TranslateIcon',
      )
      expect(iconSelectBox.props.items).toEqual(languages)
      expect(iconSelectBox.props.onSelect.name).toBe('changeLanguage')
    }

    expect(container).toMatchSnapshot()
  })

  it('Go to the root page when the app title is clicked', () => {
    const mockPush = jest.fn()
    const history = createMemoryHistory({ initialEntries: ['/posts/1'] })
    history.push = mockPush

    render(
      <Router location={history.location} navigator={history}>
        <Header />
      </Router>,
    )

    fireEvent.click(screen.getByText('Blog App'))
    expect(mockPush.mock.calls[0][0].pathname).toBe('/')
  })

  it('URL is changed by selecting language', () => {
    const mockPush = jest.fn()
    const history = createMemoryHistory({ initialEntries: ['/posts/1'] })
    history.push = mockPush

    render(
      <Router location={history.location} navigator={history}>
        <Header />
      </Router>,
    )

    {
      act(() => {
        mockIconSelectBox.mock.calls[0][0].onSelect('ja')
      })
      expect(mockPush.mock.calls[0][0].pathname).toBe('/ja/posts/1')
    }

    {
      act(() => {
        mockIconSelectBox.mock.calls[0][0].onSelect('ko')
      })
      expect(mockPush.mock.calls[1][0].pathname).toBe('/ko/posts/1')
    }

    {
      act(() => {
        mockIconSelectBox.mock.calls[0][0].onSelect('en')
      })
      expect(mockPush.mock.calls[2][0].pathname).toBe('/en/posts/1')
    }
  })

  describe('Support multiple languages', () => {
    const TestComponent = ({
      initLanguage,
    }: {
      readonly initLanguage: string
    }) => {
      const { i18n } = useTranslation()
      useEffect(() => {
        i18n.changeLanguage(initLanguage)
      }, [])
      return <Header />
    }

    it('English', () => {
      render(
        <BrowserRouter>
          <TestComponent initLanguage="en" />
        </BrowserRouter>,
      )

      expect(screen.queryByText('Blog App')).toBeInTheDocument()
      expect(screen.queryByText('ブログアプリ')).not.toBeInTheDocument()
      expect(screen.queryByText('블로그 앱')).not.toBeInTheDocument()
    })

    it('Japanese', () => {
      render(
        <BrowserRouter>
          <TestComponent initLanguage="ja" />
        </BrowserRouter>,
      )

      expect(screen.queryByText('Blog App')).not.toBeInTheDocument()
      expect(screen.queryByText('ブログアプリ')).toBeInTheDocument()
      expect(screen.queryByText('블로그 앱')).not.toBeInTheDocument()
    })

    it('Korean', () => {
      render(
        <BrowserRouter>
          <TestComponent initLanguage="ko" />
        </BrowserRouter>,
      )

      expect(screen.queryByText('Blog App')).not.toBeInTheDocument()
      expect(screen.queryByText('ブログアプリ')).not.toBeInTheDocument()
      expect(screen.queryByText('블로그 앱')).toBeInTheDocument()
    })
  })
})
