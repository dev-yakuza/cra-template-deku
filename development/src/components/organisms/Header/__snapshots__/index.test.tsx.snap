// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Rendered well 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  box-sizing: border-box;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: fixed;
  z-index: 1100;
  top: 0;
  left: auto;
  right: 0;
  background-color: #1976d2;
  color: #fff;
}

@media print {
  .emotion-0 {
    position: absolute;
  }
}

.emotion-1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  min-height: 56px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media (min-width:600px) {
  .emotion-1 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:0px) {
  @media (orientation: landscape) {
    .emotion-1 {
      min-height: 48px;
    }
  }
}

@media (min-width:600px) {
  .emotion-1 {
    min-height: 64px;
  }
}

.emotion-2 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-3 {
  color: initial;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-4 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.6;
  letter-spacing: 0.0075em;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: inherit;
}

.emotion-5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-5::-moz-focus-inner {
  border-style: none;
}

.emotion-5.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-5 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-5:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-5:hover {
    background-color: transparent;
  }
}

.emotion-5.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-7 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed emotion-0"
    style="background-color: rgb(255, 255, 255);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular emotion-1"
    >
      <div
        class="MuiBox-root emotion-2"
      >
        <a
          class="emotion-3"
          href="/"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap emotion-4"
          >
            Blog App
          </h6>
        </a>
      </div>
      <button
        aria-controls="IconSelectBox"
        aria-haspopup="true"
        aria-label="Open menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-5"
        tabindex="0"
        title="Open menu"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-6"
          data-testid="TranslateIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12.87 15.07-2.54-2.51.03-.03c1.74-1.94 2.98-4.17 3.71-6.53H17V4h-7V2H8v2H1v1.99h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11.76-2.04zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2l-4.5-12zm-2.62 7 1.62-4.33L19.12 17h-3.24z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root emotion-7"
        />
      </button>
    </div>
  </header>
</div>
`;
